### DISCUSSION TITLE
React Flow Controls Not Working (Zoom In, Zoom Out, Fit View) Kindly help

### DISCUSSION THREAD

**Discussion by saad-majeed26** [xyflow/xyflow#4901] [Q&A]
I'm encountering an issue with the React Flow Controls component. Specifically, the zoom in, zoom out, and fit view functionalities are not working. The lock control, however, works as expected. Below is the relevant code: 

```
 <Box sx={{ position: 'relative', width: '100%', height: '100%' }} onClick={handleBackgroundClick}>
      <ReactFlow
        {...flowProps}
        nodes={nodes}
        edges={edges}
        edgeTypes={edgeTypes}
        onNodesChange={isExecution ? undefined : onNodesChange}
        onEdgesChange={isExecution ? undefined : onEdgesChange}
        onConnect={isExecution ? undefined : onConnect}
        onNodeContextMenu={handleNodeContextMenu}
        nodeTypes={nodeTypes}
        onNodeClick={onNodeClick}
        onEdgeClick={handleEdgeClick}
        className="react-flow-edges-deletable"
        fitView
      >
          <Background
            variant={BackgroundVariant.Dots}
            size={2}
            gap={24}
           
          />
        <Controls/>
        <MiniMap />
        <FlowControls/>
      </ReactFlow>
 </Box>
```

**Parent Component**

`
 <ReactFlowProvider>
                  <FlowChart workflowId={workflowId || ''} />
                </ReactFlowProvider>
`

**Comment by swami1302** [comment]
The Controls component is not working for zooming and fitting the view, while the lock control works fine. A possible reason could be the fitView prop in the ReactFlow component. If `fitView` is set as a boolean (fitView), it might interfere with the built-in zoom and fit view functionalities.

1. Configure fitView with a padding value
```
<ReactFlow
  // other props
  fitView={{ padding: 0.1 }}
>
  {/* children */}
</ReactFlow>
```
2. Remove fitView entirely
```
<ReactFlow
  // other props
>
  {/* children */}
</ReactFlow>
```
Also, check for any CSS restrictions or external styles that might be affecting the Controls component. If the issue persists, reviewing the console for errors or warnings could help.

Relevant documentation:

fitView : [Prop](https://reactflow.dev/api-reference/types/fit-view-options)
Controls [Component](https://reactflow.dev/api-reference/types/fit-view-options)
Let me know if this helps or if there's more context to consider. 🚀

**Comment by yulitg2110** [comment]
I got the same issue.

**Comment by cchin25** [comment]
I'm seeing this behavior with both the Controls and Minimap. I'm on xyflow v12.8.3 with React v19

I'm using a controlled flow with Zustand + Immer, though I wouldn't think that would matter? The onClick events fire using a simple `onClick={() => console.log('hello')} but nothing's happening with pannable or zoomable.

When I pan on the actual flow, the minimap does correctly move the visualization around. There are no console errors emitted.

**Reply by swami1302** [reply]
Could you share a minimal reproducible example (or snippet) of how you’re rendering `<ReactFlow /> `along with `<Controls />` and `<MiniMap />`?

**Comment by finch-neosource** [comment]
I'm not sure if this is the same issue, but I noticed that fit to view isn't working for me when customer react nodes have zero width or height.

https://github.com/xyflow/xyflow/issues/5454
